casper/src/main/resources/Lockbox.rho

```diff
@@ -0,0 +1,56 @@
+//scalapackage coop.rchain.rholang
+
+//Registry info:
+//  sk: 78623a45fb0de7c4bc35c7099e8037c09721fff814e58d419b0e1b21f5a630bb
+//  pk: 7e59141cbc0dfb24ca0e4fe57666895190740778871744be3cec262147e02f56
+//  user == pk
+//  timestamp: 1552131653401
+//Resulting unforgable name: Unforgeable(0xbcfcd4f7ab0b4e1f6c1018f38c42d9721180b9fc92ba4bbe8f00b5010513f80e)
+//  ==> signature data == 2a3eaa013b0a0d2a0b10feffffffffffffffff010a2a5a280a243a220a20d7da018b7065def0035d962198cf1581a6aae7509aff8ca75a4035ac5a2ce49c1001
+//  ==> signature == 5af8389f2e933cc7e3e701dbf53b75ac0a3ba05265a3fb166f010cf3b301a4837351f0ec752b334620f7efd56b9971ebedd9e6c137efcefe2bfa44627d866d0c
+//URI derived from pk == `rho:id:9dsr55z1js13x346yhhecx66ns486i3yqf6jafrd9p9hdrrbxjqmyu`
+new
+  Lockbox,
+  rs(`rho:registry:insertSigned:ed25519`),
+  uriOut,
+  getDeployData(`rho:deploy:params`)
+in {
+  contract Lockbox(@"lock", @ownerPubKeyBytes, protected, ret) = {
+    new lockbox, getIdentity in {
+
+      contract getIdentity(retCh) = {
+        new ch in {
+          getDeployData ! (*ch) |
+          for (@_,@_,@userId, @_ <- ch) {
+            retCh ! (userId)
+          }
```

Comment:
This is equivalent to ethereum's msg.sender. I would prefer not to have that available everywhere in the rholang. I'm not opposed to it happening "at the door" though.

Reply:
What's wrong with msg.sender? Why would you prefer to not have it available everywhere?

I imagine the Lockbox to be a basic building block for safe, private storage of unforgable names / OCAPs. One could trivially implement a PO Box contract based on that, and/or a 'private locker' (basically a map in a lockbox).

Reply:
The trouble with `msg.sender` is that it allows users to work _around_ ocap discipline and instead implement access control lists. In ethereum it is ubiquitous to have a ACL of length one, the so-called "owner".

There are a few discussions in pub-rchain#wallet about this.

I'm not claiming this is a disastrous design decision, but I think if we make it, we should stop promoting rholang as "enforcing ocap discipline" and maybe pivot to "permitting ocap discipline".

Here's a design I proposed on discord
<blockquote>
I'm imagining a phase that happens between when the user sends new code to a validator, and when that code gets parred into the tuplespace that would be the "door" phase. The workings are not conceptually hard. Anytime code contains rho:io:stdout or rho:lockbox:<some pubkey> or whatever, the node checks to see whether the deploying pubkey (msg.sender) has access to those channels. If so the new names are instantiated to the correct unforgeable IDs. If not... I'm open to suggestions. They are just arbitrary new names? Deploy fails?
So how do we know whether a particular deployer "has access" to a given capability? If it's in the registry or their, then eveyone has access to it.  There is also a lockbox/safe/locker associated with each pubkey (implicitly empty until first use) only accessible in this "door" phase. If a capability is in the deployers locker, they have access to it.
</blockquote>

Reply:
I see. I think we should revisit this before mainnet, it might make sense to stick to ocap-only. It's a bit less user friendly though (it's easier to leak/lose track of an ocap than a public key).

Funny though how this was introduced to implement safe ocap storage in the first place (a lockbox is a safe/locker of size 1, and when discussing this change I had a safe in mind as well).